### ISSUE TITLE
static fields don't work with filter_grad(arg=...)?

### ISSUE THREAD

**Issue by ywrt** [patrick-kidger/equinox#128]
If I define a class with a static field (via "field: int = eqx.static_field") and then attempt to use jax.tree_map to make a gradient mask for eqx.filter_grad, it will error with "ValueError: `filter_spec` must consist of booleans and callables only."

There seems to be an inconsistency, where the tree_map/partition/filter are sure that the static field doesn't exist (it's never visible as a leaf), but filter_grad sees it and is unhappy.

This makes it rather difficult to construct a pytree that can be passed into filter_grad(arg=)!

Any ideas on how I should solve this?

**Comment by ywrt** [comment]
Something very weird is happening. I have two colab notebooks side by side:
The code below runs fine in one of them, and crashes in the other with: "ValueError: `filter_spec` must consist of booleans and callables only."

If I remove the jit, it doesn't crash. If I add it back in, it does crash. Note that the commented out call to 'eqx.partition' crashes with and without the jit!

class MWE(eqx.Module):
  w: jnp.ndarray
  def __call__(self, x, loss=0.):
    return x * self.w, 0.
def lfn(m, x):
  p,_ = m(x)  
  return jnp.sum(p), 1.

model = MWE( w=jnp.ones((1,)))
gmask = jax.tree_map(lambda x : True, model)

#a,b = eqx.partition(model, gmask)

vg = eqx.filter_value_and_grad(lfn, has_aux=True, arg=gmask)
loss, xtra = jax.jit(vg)(model, jnp.array([1]))

**Comment by ywrt** [comment]
I think I'm having a bad day!

Without having changed anything, now both notebooks work fine and I can't make either of them crash.
But my more complex example still crashes!

Poking further, in my existing code, if I pass in grad_mask from outside the jit, it crashes. (using "jax.tree_map(lambda x : True, model)" to construct)
If I construct grad_mask inside the jit with the same tree_map, it doesn't crash.

There is a small difference: If I print out the value of grad_mask when I pass it in externally, it shows
<pre>
    ...
   ANormedEmbed(w=bool[]),
   ...
</pre>
but if I construct it inside the jit it shows:
<pre>
    ...
    ANormedEmbed(w=True),
   ...
</pre>
I'm not sure how much I should care about this? it's obviously only during the tracing pass, so ... ??

**Comment by ywrt** [comment]
Ok, I see the issue now. It seems that eqx.partition doesn't work under jit tracing and it's a heisenbug due to the myriad layers of caching. 
 
This code reliably crashes in a fresh instance:
<pre>
class MWE(eqx.Module):
  w: jnp.ndarray
  
model = MWE( w=jnp.ones((1,)))
gmask = jax.tree_map(lambda x : True, model)

a,b = jax.jit(eqx.partition)(model, gmask)
</pre>
with
<pre>
ValueError: `filter_spec` must consist of booleans and callables only.
</pre>

**Comment by patrick-kidger** [comment]
This is expected: the mask is being traced, and so `filter_spec` does not contain booleans or callables. (In this case, it contains a tracer.)

Your second-to-last message captures the essential point: you are tracing the filter specification. However the point of the filter specification is to determine the structure of the XLA graph you'd like to construct. Tracers themselves are very different: these are inputs to the XLA graph.

In short, don't trace the filter specification (i.e. don't pass it as a non-static argument across a JIT boundary) and you should find that things work fine.

**Comment by ywrt** [comment]
I re-wrote my code to pull the gradient mask outside the jit and everything is good. Thank you very much :)
